### PR TITLE
Constraint system updates

### PR DESCRIPTION
**Pull request by ETatuzova**

### CODE REVIEW

File: crypto3/libs/blueprint/include/nil/blueprint/blueprint/plonk/assignment.hpp [hunk 1]
```diff
@@ -152,9 +152,9 @@ namespace nil {
             using ranges = std::vector<std::pair<std::size_t, std::size_t>>;
 
         public:
-            static constexpr const std::size_t private_storage_index = std::numeric_limits<std::size_t>::max();
-            static constexpr const std::size_t batch_private_storage_index = std::numeric_limits<std::size_t>::max() - 1;
-            static constexpr const std::size_t batch_constant_storage_index = std::numeric_limits<std::size_t>::max() - 2;
+            static constexpr const std::int64_t private_storage_index = std::numeric_limits<std::size_t>::max();
```
Comment (ETatuzova) [inline crypto3/libs/blueprint/include/nil/blueprint/blueprint/plonk/assignment.hpp:155]: Change it and check according to signed index type

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/arithmetization/plonk/constraint_system.hpp [hunk 2]
```diff
@@ -50,8 +50,10 @@ namespace nil {
     namespace crypto3 {
         namespace zk {
             namespace snark {
-
                 /************************* PLONK constraint system ****************************/
+                constexpr static int const PLONK_SPECIAL_SELECTOR_ALL_USABLE_ROWS_SELECTED = -1;
```
Comment (martun) [inline crypto3/libs/zk/include/nil/crypto3/zk/snark/arithmetization/plonk/constraint_system.hpp:54]: Let's try to use std::numeric_limits<size_t>::max - 1, I'm afraid of negative values, they may get converted to unsigend somewhere by a mistake.

File: parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 3]
```diff
@@ -468,14 +469,30 @@ namespace nil {
...
+                            else{
+                                switch( l_table.tag_index ){
+                                case PLONK_SPECIAL_SELECTOR_ALL_USABLE_ROWS_SELECTED:
+                                    lookup_tag = mask_assignment;
+                                    break;
+                                case PLONK_SPECIAL_SELECTOR_ALL_NON_FIRST_USABLE_ROWS_SELECTED:
+                                    lookup_tag = mask_assignment - lagrange0;
+                                    break;
+                                default:
+                                    BOOST_ASSERT_MSG(false, "Unknown special selector");
```
Comment (martun) [inline parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:493]: Maybe move the general case into the "default"? lookup_tag = plonk_columns.selector(l_table.tag_index);

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 4]
```diff
@@ -6,6 +6,7 @@
 // Copyright (c) 2023 Elena Tatuzova <e.tatuzova@nil.foundation>
 // Copyright (c) 2023 Martun Karapetyan <martun@nil.foundation>
 // Copyright (c) 2024 Vasiliy Olekhov <vasiliy.olekhov@nil.foundation>
+// Copyright (c) 2024 Elena Tatuzova <e.tatuzova@nil.foundation>
```
Comment (martun) [inline crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:9]: You already have this.

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 5]
```diff
@@ -589,10 +624,7 @@ namespace nil {
                         std::unordered_map<typename FieldType::value_type, std::size_t> sorting_map;
                         for (std::size_t i = 0; i < reduced_value.size(); i++) {
                             for (std::size_t j = 0; j < usable_rows_amount; j++) {
-                                if(sorting_map.find(reduced_value[i][j]) != sorting_map.end())
-                                    sorting_map[reduced_value[i][j]]++;
-                                else
-                                    sorting_map[reduced_value[i][j]] = 1;
+                                sorting_map[reduced_value[i][j]] = 1;
```
Comment (martun) [inline crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:627]: Are we sure the values in reduced_value don't repeat? Maybe we keep something like:

 if(sorting_map.find(reduced_value[i][j]) != sorting_map.end())
                                  throw std::invalid_argument("Values in sorting map must not repeat");
                              else
                                  sorting_map[reduced_value[i][j]] = 1;

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp [hunk 6]
```diff
@@ -745,8 +763,9 @@ namespace nil {
                         std::vector<typename FieldType::value_type> lookup_input;
                         for( std::size_t g_id = 0; g_id < lookup_gates.size(); g_id++ ){
                             const auto &gate = lookup_gates[g_id];
+                        typename FieldType::value_type selector_value;
                             auto key = std::tuple(gate.tag_index, 0, plonk_variable<typename FieldType::value_type>::column_type::selector);
-                            auto selector_value = evaluations[key];
+                            selector_value = evaluations[key];
```
Comment (martun) [inline crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:768]: If you want to explicitly specify the type, you can do it here:

typename FieldType::value_type selector_value = evaluations[key];

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/verifier.hpp [hunk 7]
```diff
@@ -330,6 +330,31 @@ namespace nil {
...
+                        typename FieldType::value_type mask_value = FieldType::value_type::one() -
+                            proof.eval_proof.eval_proof.z.get(FIXED_VALUES_BATCH, common_data.permuted_columns.size() * 2, 0) -
+                            proof.eval_proof.eval_proof.z.get(FIXED_VALUES_BATCH, common_data.permuted_columns.size() * 2 + 1, 0);
+                        typename FieldType::value_type shifted_mask_value = FieldType::value_type::one() -
+                            proof.eval_proof.eval_proof.z.get(FIXED_VALUES_BATCH, common_data.permuted_columns.size() * 2, 1) -
+                            proof.eval_proof.eval_proof.z.get(FIXED_VALUES_BATCH, common_data.permuted_columns.size() * 2 + 1, 1);
+
+                        // All rows selector
+                        {
+                            auto key = std::make_tuple( -1, 0, plonk_variable<typename FieldType::value_type>::column_type::selector);
```
Comment (martun) [inline crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/verifier.hpp:342]: Can you create a const variable for value -1, just like you did in gate argument with PLONK_SPECIAL_SELECTOR_ALL_USABLE_ROWS_SELECTED. Maybe we need to do something like

static const int PLONK_SPECIAL_SELECTOR_ALL_USABLE_ROWS_SELECTOR_INDEX = -1;

File: parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/gates_argument.hpp [hunk 8]
```diff
@@ -108,15 +110,29 @@ namespace nil {
...
                                 const variable_type& var = variables[i];
 
                                 // Convert the variable to polynomial_dfs variable type.
                                 polynomial_dfs_variable_type var_dfs(var.index, var.rotation, var.relative,
                                     static_cast<typename polynomial_dfs_variable_type::column_type>(
                                         static_cast<std::uint8_t>(var.type)));
 
-                                polynomial_dfs_type assignment = assignments.get_variable_value(var_dfs, domain);
+                                polynomial_dfs_type assignment;
+                                if( var.index >= 0){
```
Comment (martun) [inline parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/gates_argument.hpp:122]: This code needs to get into function get_variable_value itself. We don't even check here that it's a selector, before returning the special values.

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp (hunk 5 above)
Comment (ETatuzova) [reply crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp:627]: Values in `reduced_value` may repeat.
Initially `sorting_map` contains number of each occurencies in `lookup_input` + 1.

When we meet some `v` in `reduced_value` for the first time during sorting procedure we pusht one occurrence from `reduced_value` and all from `reduced_input` to sorted columns.

Next times we'll push only one value from `reduced_value` and nothing from `reduced_input`. That's why we set sorting_map[value] = 1 after first occurence.

So, `sorting_map`  containt information -- how many times `v` should be pushed to sorted columns next time

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/arithmetization/plonk/constraint_system.hpp (hunk 2 above)
Comment (ETatuzova) [reply crypto3/libs/zk/include/nil/crypto3/zk/snark/arithmetization/plonk/constraint_system.hpp]: Done in last commit

File: parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp (hunk 3 above)
Comment (ETatuzova) [reply parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp]: Replaced switch by direct if-elses

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/verifier.hpp (hunk 7 above)
Comment (ETatuzova) [reply crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/verifier.hpp]: Done. Now constants are used in this place

File: parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/gates_argument.hpp (hunk 8 above)
Comment (ETatuzova) [reply parallel-crypto3/libs/parallel-zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/gates_argument.hpp]: Now everywhere we check that var.type == selector

File: crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp (hunk 4 above)
Comment (ETatuzova) [reply crypto3/libs/zk/include/nil/crypto3/zk/snark/systems/plonk/placeholder/lookup_argument.hpp]: Sorry:)

**Review by Iluvmagick — APPROVED** [review]

**Review by martun — APPROVED** [review]

**Pull request merged**